### PR TITLE
i#5383 mac a64, part 2: Fix compiler warnings

### PR DESCRIPTION
**Pull request by derekbruening**
Fixes many unused-variable warnings when building the core and
extension libraries.  This enables a config without tests or docs to
build.  The drsyms libraries are missing so clients/ is not yet built.

Issue: #5383

### CODE REVIEW

File: core/arch/interp.c [hunk 1]
```diff
@@ -4656,7 +4656,7 @@ static void
 build_native_exec_bb(dcontext_t *dcontext, build_bb_t *bb)
 {
     instr_t *in;
-    opnd_t jmp_tgt;
+    opnd_t jmp_tgt IF_AARCH64(UNUSED);
```
Comment (abhinav92003) [inline core/arch/interp.c:4659]: Isn't this used on L4748 below for AArch64?

```
jmp_tgt = opnd_create_pc(bb->start_pc);
```

File: core/arch/interp.c [hunk 2]
```diff
@@ -7247,7 +7248,9 @@ decode_fragment(dcontext_t *dcontext, fragment_t *f, byte *buf, /*IN/OUT*/ uint
     bool coarse_elided_ubrs = false;
     dr_isa_mode_t old_mode;
     /* for decoding and get_ibl routines we need the dcontext mode set */
-    bool ok = dr_set_isa_mode(dcontext, FRAG_ISA_MODE(f->flags), &old_mode);
+    DEBUG_DECLARE(bool ok =)
```
Comment (abhinav92003) [inline core/arch/interp.c:7251]: Maybe we should have a macro for such patterns where the returned boolean is supposed to be checked only in debug mode.

File: core/ir/aarch64/codec.c [hunk 3]
```diff
@@ -1323,6 +1323,7 @@ encode_opnd_b_const_sz(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *en
     return false;
 }
 
+#if 0  /* Currently unused. */
 /* h_const_sz: Operand size for half (16-bit) vector elements
```
Comment (abhinav92003) [inline core/ir/aarch64/codec.c:1327]: If these routines are not exported, isn't it fine to remove them? Or do we think they might be used in the near future?

File: core/ir/instr_shared.c [hunk 4]
```diff
@@ -1490,15 +1490,15 @@ instr_decode_cti(dcontext_t *dcontext, instr_t *instr)
      */
     if (!instr_opcode_valid(instr) ||
         (instr_is_cti(instr) && !instr_operands_valid(instr))) {
-        byte *next_pc;
+        DEBUG_EXT_DECLARE(byte * next_pc;)
```
Comment (abhinav92003) [inline core/ir/instr_shared.c:1493]: DEBUG_EXT_DECLARE doesn't seem to do anything that's "declaration" related. Why not just IF_DEBUG/DODEBUG?

File: core/utils.c [hunk 5]
```diff
@@ -1864,12 +1864,9 @@ d_r_notify(syslog_event_type_t priority, bool internal, bool synch,
            const char *fmt, ...)
 {
     char msgbuf[MAX_LOG_LENGTH];
-    int size;
     va_list ap;
     va_start(ap, fmt);
-    /* FIXME : the vsnprintf call is not needed in the most common case where
```
Comment (abhinav92003) [inline core/utils.c:1870]: This fixme is fixed now?

File: core/utils.h [hunk 6]
```diff
@@ -1567,15 +1567,21 @@ enum { LONGJMP_EXCEPTION = 1 };
...
 /* Use to shut up the compiler about an unused variable when the
  * alternative is a painful modification of more src code. Our
  * standard is to use this macro just after the variable is declared
  * and to use it judiciously.
  */
-#define UNUSED_VARIABLE(pv)           \
-    {                                 \
-        void *unused_pv = (void *)pv; \
-        unused_pv = NULL;             \
+#define UNUSED_VARIABLE(pv)                  \
```
Comment (abhinav92003) [inline core/utils.h:1581]: Why do we need UNUSED_VARIABLE and UNUSED both?

Edit: Oh okay. We use the former to not have to put too many pre-processor defines on the declaration. Whereas the latter is used when the variable is unused in all or only a small number of environments.

File: core/unix/os.c [hunk 7]
```diff
@@ -4702,7 +4699,7 @@ void
 os_syslog(syslog_event_type_t priority, uint message_id, uint substitutions_num,
           va_list args)
 {
-    int native_priority;
+    int native_priority UNUSED;
```
Comment (abhinav92003) [inline core/unix/os.c:4702]: What does the native_priority var here achieve?

File: core/unix/os.c [hunk 8]
```diff
@@ -6039,7 +6035,7 @@ handle_execve(dcontext_t *dcontext)
      */
     char *fname;
     bool x64 = IF_X64_ELSE(true, false);
-    bool expect_to_fail = false;
+    bool expect_to_fail UNUSED = false;
```
Comment (abhinav92003) [inline core/unix/os.c:6038]: But it's used on L6151:
```
if (should_inject && DYNAMO_OPTION(early_inject) && !expect_to_fail) {
```


Is unused a no-op if it is actually used?

**Review by abhinav92003 — APPROVED** [review]

File: core/utils.h (hunk 6 above)
Comment (abhinav92003) [reply core/utils.h:1581]: s/only a small number of environments/simple set of environments/

File: core/arch/interp.c (hunk 1 above)
Comment (derekbruening) [reply core/arch/interp.c:4659]: It is only set there: never read after that for a64.  This clang 14 is very aggressive at these warnings: as you see it found tons of such vars.

File: core/ir/aarch64/codec.c (hunk 3 above)
Comment (derekbruening) [reply core/ir/aarch64/codec.c:1327]: Let's ask the owners: @AssadHashmi the latest clang (14) flags a bunch of codec.c functions as unused -- were these used in the past and are now truly dead, or  are these planned to be used in the future?

File: core/ir/instr_shared.c (hunk 4 above)
Comment (derekbruening) [reply core/ir/instr_shared.c:1493]: I thought DEBUG_EXT_DECLARE was for CLIENT_ASSERT which is subtly different in which places it's enabled vs DEBUG -- at least that's what the comment says:
```C
/* For use in CLIENT_ASSERT or elsewhere that exists even if
 * STANDALONE_DECODER is defined.
 */
```

File: core/utils.h (hunk 6 above)
Comment (derekbruening) [reply core/utils.h:1581]: Also I think we didn't find a cross-platform compiler marking that worked well in the past: we were avoiding pragmas in VS

File: core/unix/os.c (hunk 7 above)
Comment (derekbruening) [reply core/unix/os.c:4702]: Didn't want to deal with the unfinished code: you're asking to delete the function body code that just sets native_priority and never does anything with it?

File: core/unix/os.c (hunk 8 above)
Comment (derekbruening) [reply core/unix/os.c:6038]: Used only for LINUX, not OSX.  Yes, it's a nop if it's actually used, so doesn't hurt LINUX build.

File: core/arch/interp.c (hunk 2 above)
Comment (derekbruening) [reply core/arch/interp.c:7251]: Could be confusing though if you had:
```C
DEBUG_CHECK_RET_TRUE(dr_set_isa_mode(dcontext, FRAG_ISA_MODE(f->flags), &old_mode));
```
Which the reader will probably think is only executed at all in debug build.
Would have to name it ALWAYS_RUN_BUT_CHECK_RETVAL_IN_DEBUG() or sthg?

File: core/ir/aarch64/codec.c (hunk 3 above)
Comment (derekbruening) [reply core/ir/aarch64/codec.c:1327]: No reply yet: I guess they can always be restored from version control so I will remove them.

File: core/arch/interp.c (hunk 2 above)
Comment (derekbruening) [reply core/arch/interp.c:7251]: Just going to leave the code as-is.

File: core/ir/instr_shared.c (hunk 4 above)
Comment (derekbruening) [reply core/ir/instr_shared.c:1493]: Sticking with what is intended for use in this situation.

File: core/unix/os.c (hunk 7 above)
Comment (derekbruening) [reply core/unix/os.c:4702]: Going to leave it.

File: core/ir/aarch64/codec.c (hunk 3 above)
Comment (derekbruening) [reply core/ir/aarch64/codec.c]: Removing them breaks the aarch64_check_codec_order.py script which statically examines this code and compares to opnd_defs.txt.  So we'd have to remove lines from opnd_defs.txt too.  I'm going to put them back and defer to @AssadHashmi on a longer-term fix.

**Pull request merged**